### ISSUE TITLE
Config validation mechanism is incomplete

### ISSUE THREAD

**Issue by marschwar** [detekt/detekt#3935]
## Expected Behavior
Configuration errors (like type mismatches) should be detected at build time (for the detekt-default-config.yml) or during the validation phase of the detekt run.

## Observed Behavior
Detekt does provide a mechanism to validate the configuration via `ConfigValidator` but the default implementation only checks if the properties used in the users config match those of the `detekt-default-config.yml`. What is missing:

* Validation that the value that is provided matches the expected type in the rule. -> Error is reported when the rule config property is evaluated at runtime. Most likely the message is rather cryptic.
* Validation of the types in the `detekt-default-config.yml` -> If the property in question is overwritten in our `config/detekt/detekt.yml` we might not recognize the issue and end up shipping this faulty config and might end up breaking user's builds.

## Steps to Reproduce
Change the `config/detekt/detekt.yml`

```yaml
  # from
  MagicNumber:
    ignoreNumbers: ['-1', '0', '1', '2', '100', '1000']

  # to
  MagicNumber:
    ignoreNumbers: [-1, 0, 1, 2, 100, 1000]
```

And run
```
java -jar ./build/detekt-cli-all.jar \
  --build-upon-default-config \
  --input detekt-api \
  --config config/detekt/detekt.yml
```

**Comment by BraisGabin** [comment]
I think that approach similar to #4104 could help. Another option is to use an annotation processor now that KSP is stable.

I think that this could be implemented along side with #3242.

My idea: instead of create a single yml with all the configurations we can create a file for each rule module. And, when the user asks to generate the configuration file the cli/gradle plugin will concat all those files and generate the desired file. We should concat a header with the other configurations but that should be easy.
